### ISSUE TITLE
proposal: `require_trailing_commas` should also flag arrays

### ISSUE THREAD

**Issue by alexeyinkin** [dart-lang/linter#3607]
## Good Examples 

```dart
final list = [
  'longer text so that this definitely does not fit in one line',
  2,  // <-- comma here
];
```

## Bad Examples

```dart
final list = [
  'longer text so that this definitely does not fit in one line',
  2
];
```

## Discussion

`require_trailing_commas` only flags missing commas in argument lists but not in array construction. I think it should flag arrays as well because the point of a trailing comma is to allow adding items without affecting the line with the last item, and this should clearly include arrays.

The real-world examples include `children` list in Flutter where widgets very often do not fit on one line.

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [ ] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [ ] If there's any prior art (e.g., in other linters), please add references here.
- [ ] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn’t any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

**Comment by srawlins** [comment]
+1; I think it makes sense and `require_trailing_commas` is "experimental" so now is the time to make the change.

**Comment by bwilkerson** [comment]
@pq Do you remember the genesis of this lint? My only concern here is that this might not have been the intent of the original request, so we might be breaking someone by changing it.

I suppose we could always add a `require_trailing_commas_in_collections` rule that operates on list, map, and set literals.

**Comment by pq** [comment]
IIRC this was upstreamed from an internal lint.  @lauweijie may have thoughts here.

**Comment by lauweijie** [comment]
Yes, this was indeed upstreamed from an internal lint and the motivation was for better readability after dartfmt.

Given the flutter [style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-a-trailing-comma-for-arguments-parameters-and-list-items-but-only-if-they-each-have-their-own-line) also includes list items, this sounds like a reasonable suggestion. Whether to update the existing `require_trailing_commas` lint or have a separate one as @bwilkerson suggested is open for discussion.

**Comment by alexeyinkin** [comment]
There is also this motion for more trailing commas: https://github.com/dart-lang/language/issues/1402
Those cases are more controversial, and will likely require separate lints if implemented. So we should be prepared that we cannot fit all trailing commas in a single lint.

**Comment by a14n** [comment]
FWIW #3340 is open for a while and should fix this issue.

**Comment by a14n** [comment]
This issue can be closed as #3650 is now merged.
